Porting Collections Off of Tower CLI

beeankha opened the pull request:
##### SUMMARY

The related AWX issue can be found here: https://github.com/ansible/awx/issues/5145.  This is a collaborative PR between myself and @john-westcott-iv . 


##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - Collections

john-westcott-iv committed:
Initial implementation of Pull #5337

john-westcott-iv committed:
Attempting to fix logout loop

john-westcott-iv committed:
Only warn if we can't release a tower token

john-westcott-iv committed:
Auto-handle check_mode on post, patch, delete and update_if_needed methods

john-westcott-iv committed:
Logout is now handled by exit_json

john-westcott-iv committed:
Fix fail_json and remove redundant handle check mode

john-westcott-iv committed:
Converted tower_credential_type.py

john-westcott-iv committed:
Converted tower_organization

john-westcott-iv committed:
Removing default_check_mode

john-westcott-iv committed:
Converted tower_user

john-westcott-iv committed:
Converted tower_settings

Changed comments

john-westcott-iv committed:
Adding team_fields

Convert job_list and inventory modules, other changes to make sanity tests pass

squidboylan commented on awx_collection/plugins/modules/tower_license.py:
```diff
@@ -0,0 +1,88 @@
+#!/usr/bin/python
+# coding: utf-8 -*-
+
+# (c) 20189, John Westcott IV <john.westcott.iv@redhat.com>
```
Typo in the copyright year

squidboylan commented on awx_collection/plugins/modules/tower_license.py:
```diff
@@ -0,0 +1,88 @@
+    json_output = {'changed': False}
+
+    if not module.params.get('eula_accepted'):
+        module.fail_json(msg='You must accept the EULA by passing in the param eula_acepte as True')
```
typo in the help text, should be `eula_accepted` not `eula_acepte`

AlanCoding commented on awx_collection/plugins/module_utils/tower_api.py:
```diff
@@ -0,0 +1,372 @@
+            existing_field = existing_item.get(field, None)
+            new_field = new_item.get(field, None)
+            # If the two items don't match and we are not comparing '' to None
+            if existing_field != new_field and not (existing_field in (None, '') and new_field == ''):
```
Let me backref where the same things are done in tower-cli

https://github.com/ansible/tower-cli/blob/b3d49ce2cf4ff175da4fb5ed1bdef110a7c81083/tower_cli/models/base.py#L401

We need to think of what this loses compared to the old logic. One thing is the null-like fields. I suspect we may have a problem with the `limit` field specifically, where it needs to be set to empty string. We should be able to skip null values, but I don't think we can for empty string values.

AlanCoding commented on awx_collection/plugins/module_utils/tower_api.py:
```diff
@@ -0,0 +1,372 @@
+                if not handle_response:
+                    return response
+                elif response['status_code'] == 200:
+                    existing_return['changed'] = True
```
link https://github.com/ansible/awx/issues/5474

squidboylan commented on awx_collection/plugins/module_utils/tower_api.py:
```diff
@@ -0,0 +1,372 @@
+        super(TowerModule, self).__init__(argument_spec=args, **kwargs)
+
+        # If we have a tower config, load it
+        if self.params.get('tower_config_file'):
```
The tower cli has a few default config file locations https://docs.ansible.com/ansible-tower/3.5.3/html/towerapi/tower_cli.html#file-locations , we probably need to be backwards compatible with that

john-westcott-iv replied:
This should be resolved now. We need to see if we can somehow clone the existing tower-cli/tests/test_config.py file.

squidboylan reviewed (commented):
I left an inline comment about tower_cli.cfg default paths, another thing to note is that the tower_cli.cfg supports both yaml and `key=value` format, which we should make a decision on whether we need/want to support.

beeankha committed:
Further module conversion changes, unit test changes

Multiple module changes

Added on_change callback

Added head_endpoint

Added additional error returns

Respond with a try an ID message if multiple assets found by name via return_none_on_404 kwarg

Diferentiated between login and logout token errors

Added is_job_done method

beeankha commented on awx_collection/plugins/modules/tower_credential_type.py:
```diff
@@ -64,12 +64,11 @@
       choices: ["present", "absent"]
       type: str
-    validate_certs:
+    tower_oauthtoken:
```
Adding the `tower_oauthtoken` docstring to each converted module until everything has changed over, at which point this docstring can just be added to the `awx_collection/plugins/doc_fragments/auth.py` file instead.

squidboylan replied:
I'm not sure we want to remove the "validate_certs" option here, that is to disable validating the tower ssl certs, it seems that the oauth token option does not replace this but is instead supplementary

john-westcott-iv replied:
The validate_certs option can be removed from the individual modules because it is defined in tower_api.py line 42.

john-westcott-iv committed:
Converted tower_project

AlanCoding commented on awx_collection/plugins/modules/tower_project.py:
```diff
@@ -133,107 +143,129 @@
+            result = module.get_endpoint('/project_updates/{}/'.format(last_request['summary_fields']['current_update']['id']))['json']
+
+            if module.is_job_done(result['status']):
+                running = False
```
This should have at least a little delay in the loop. Otherwise, it'll occupy a uWSGI worker almost constantly.

AlanCoding commented on awx_collection/plugins/module_utils/tower_api.py:
```diff
@@ -0,0 +1,408 @@
+            except(Exception) as excinfo:
+                self.fail_json(changed=False, msg='Failed check mode: {0}'.format(excinfo))
+
+    def update_if_needed(self, existing_item, new_item, handle_response=True, **existing_return):
```
Before this gets too far into the weeds, I want to ask about the highest priority design item, which is the consistency of how these methods are called.

It would be really good to distinguish all methods that terminate execution - that is they do exit_json or fail_json. I would also like to be very careful that they either always terminate or never terminate. You've done an excellent job fitting a solution to do that with the on_changed callback. Whether they terminate or not is so important, I almost think it should be in the method name.

On to the call arguments, the core CRUD actions involve `update_if_needed`, `post_endpoint`, and `delete_endpoint`. Consider the arguments for post_endpoint versus this method here.

```python
def post_endpoint(self, endpoint, handle_return=True, item_type='item', item_name='', *args, **kwargs):
```

This just doesn't look the same at all. My position is that these should all have a common set of functionality - because they are terminating methods.

Likewise, for deletion, you handle the case of changed=False in the module code very elegantly, but that's not consistent with this. If we have a method `update_if_needed`, then we should have a `delete_if_needed` method so that we know that (a) it will terminate execution and (b) it could terminate with multiple types of statuses.

I come back around to the design that tower-cli used, maybe we should have a single method for CRU and another one for D. Since the module itself deals with the `state` parameter, this makes a lot of sense with the surface area that Ansible requires.

john-westcott-iv replied:
Great feed back.

In thinking about, I'm learning towards having C, U, and D methods. In the form of:
`update_if_needed`
`create_if_needed`
`delete_if_needed`

I don't want to put `_and_exit` on the functions since you can override the exit with `handle_response=False` and my gut is saying we will find some use case where we might have a "compound" action we need to take. I.e. create item 'a' and then create item 'b' which has a reference to 'a'. So the first one would handle_exit=False and the second one is True.

Each of the methods above can have a corresponding hook for `on_create`, `on_update`, and `on_delete`. Do you think those should be "registered" like I did or just make them parameters to the if_needed methods?

The R will continue to be handled via get_one and get_all (I'm thinking we can rename `get_all_endpoint` to just `get_all` for consistency).

In addition, I think it would make sense to merged CU method called `create_or_update_if_needed` so that modules can get into a very simple pattern like:
```
if state == 'absent':
    delete_if_needed(<args>)
else:
    create_or_update_if_needed(<args>)
```
Another other option would be to make a single method like: `put_object_in_state(old_object, new_object, state, ...)` which would handle all of the CUD operations in one. Any thoughts on that? That would put even the conditional logic into the module_util so the modules would only need to:
  * resolve associated IDs
  * build the object data structures
  * define any on_* methods 
  * make the put_object_in_state call.

AlanCoding replied:
I do feel partial to the CU, and D methods, exactly how your proposed code snippet is. But as long as we have consistent args with calls to `_if_needed` methods, then I give my 👍 

> my gut is saying we will find some use case where we might have a "compound" action we need to take. I.e. create item 'a' and then create item 'b' which has a reference to 'a'. So the first one would handle_exit=False and the second one is True.

Good point, we should be prepared to complex multi-action scenarios. Given that, `handle_exit=False` sounds like a good solution to me... probably better than the callback approach. If the actions got complex, then we would wind up in callback hell. The way `tower_project` waits for update could be re-written to use a `handle_exit=False` kwarg and not the callback.

john-westcott-iv replied:
Check it out now. I left the option for a callback for "simple" things, like waiting for a project update to complete. For more complex things you can give it handle_response=False but then you are on the hook for processing the return and follow up actions.

AlanCoding replied:
I'm on board with the call pattern in the modules now. Having both the callback and handle_response kwarg has a pretty good justification.

AlanCoding commented on awx_collection/plugins/modules/tower_team.py:
```diff
@@ -57,57 +67,60 @@
+    else:
+        # If the state was present and we had a team we can see if we need to update it
+        # This will return on its own
+        module.update_if_needed(team, team_fields)
```
As a specific example, following up to my last comment, both `delete_endpoint` and `post_endpoint` need the plural name used in the list endpoint, the string `"teams"` here. The method `update_if_needed` avoids this by using the URL in the data, but it doesn't need to. It could also do a PATCH to `'teams/{0}'.format(team['id'])`, and I would prefer this because it's more consistent.

The arg `item_name` doesn't seem to me like something we need to keep. We could default to looking for name, and then have some other means of overriding that on initialization of the class for the special cases.

The data is clearly needed in all cases.

So this points us in the direction of a call pattern like `'teams', team_fields`, and you could share that pattern between all the methods from module_utils.

john-westcott-iv committed:
More tower-cli-ish parsing of config files

Clear up test failures/linting errors, update unit test

Update module_utils for linter, add wait time to project module

beeankha committed:
Fix superclass syntax for < Python3

AlanCoding commented on awx_collection/test/awx/conftest.py:
```diff
@@ -111,6 +111,86 @@ def mock_load_params(self):
+        result = json.loads(module_stdout)
+        return result
+
+    return rf
```
since the code diff with the other fixture is so small, I would like to combine the two before merging this

john-westcott-iv committed:
Standardizing CRUD methods

Fixing linting and conversion issues

AlanCoding commented on awx_collection/plugins/modules/tower_team.py:
```diff
@@ -57,57 +67,53 @@
+        module.delete_if_needed(team)
+    elif state == 'present':
+        # If the state was present we can let the module build or update the existing team, this will return on its own
+        module.create_or_update_if_needed(team, team_fields, item_type='team'})
```
This looks great! I can't wait to look into check_mode and diff_mode later on. That will be so painless by comparison.

AlanCoding commented on awx_collection/plugins/module_utils/tower_api.py:
```diff
@@ -0,0 +1,534 @@
+        if existing_item:
+            return self.update_if_needed(existing_item, new_item, handle_response=handle_response, on_update=on_update)
+        else:
+            return self.create_if_needed(existing_item, new_item, handle_response=handle_response, on_create=on_create, item_type=item_type)
```
Doesn't this need to pass in the plural endpoint? So I would expect that `create_or_update_if_needed` will also need to take the plural endpoint.

Alternatively, it would make sense to me to pass that into `__init__`. It's a general property of a resource, as long as we're only declaring it once for each module, that's about the most efficient solution we could hope for. You could even take it in optionally, and just add an "s" if it's not given.

john-westcott-iv replied:
You are correct, this will need the endpoint passed into it and then on down to create_if_needed. 

The __init__ option would be cool but if we are trying to keep complex use cases in mind that would require an override option anyway. 

Finally for pluralization, I wish we could auto-pluralize a friendly name but inventorys is not valid also; if anyone ever wanted to hit /me, /config, ping, etc auto-pluralization logic could become tricky and something that would need to be maintained. It feels like it might almost work better to do an un-pluralization to derive the friendly name from the endpoint.

But I'll go ahead and get the pass through of the endpoint working and then touchup the converted modules to use the new syntax; thanks for the review.

john-westcott-iv committed:
Updating call to create_or_update_if_needed

john-westcott-iv committed:
Bug fixes for username and delete data

john-westcott-iv committed:
Modified modules to use new tower_api format

Fixed variable name typo

john-westcott-iv committed:
Updating tower_org to use the new tower_api format

Pass sanity and unit tests, update tests

Remove placeholder test function, convert tower_host module, fix misc typos

beeankha committed:
Remove JSONDecodeError exception, fix tower_host variable issue

squidboylan commented on awx_collection/plugins/module_utils/tower_api.py:
```diff
@@ -0,0 +1,552 @@
+        for config_file in config_files:
+            try:
+                self.load_config(config_file)
+            except ConfigFileException:
```
`tower-cli` throws an error if one of the default files exists but is not formatted properly, I think that behavior is better as it helps with debugging bad configs.

beeankha committed:
Change config file loading function, add py2 and py3 compatibility

john-westcott-iv committed:
Converted tower_group

Splitting out tower_inventory_source from tower_group

Copy/Paste typo fix and README update for breaking backwards compatability

Update credential_type module and unit tests

AlanCoding commented on awx_collection/plugins/modules/tower_group.py:
```diff
@@ -41,49 +46,6 @@
-      description:
-        - Refresh inventory data from its source each time a job is run.
-      type: bool
-      default: 'no'
```
We should mark this as a backward-incompatible change, in the README.md we're just making a rolling list.

john-westcott-iv replied:
Done

AlanCoding commented on awx_collection/plugins/modules/tower_group.py:
```diff
@@ -104,87 +66,62 @@
+    # Attempt to lookup the related items the user specified (these will fail the module if not found)
+    inventory_id = module.resolve_name_to_id('inventories', inventory)
+
+    # Attempt to lookup team based on the provided name and org ID
```
comment wording from copy+paste

I'd prefer to lose the comment completely

john-westcott-iv replied:
Good catch, typo fixed for now. It would be nice if we at least got rid of the 'team' in the comment and made it something more generic. My goal with the (excessive?) comments was to provide a clear template to anyone who might want to help us in this endeavor of module creation and maintenance. Since I can't tell which module someone might look at I figured lets just put them in all of them. (otherwise fixed)

AlanCoding commented on awx_collection/plugins/modules/tower_group.py:
```diff
@@ -104,87 +66,62 @@
+
+    # If the variables were spacified as a file, load them
+    if variables:
+        variables = module.load_variables_if_file_specified(variables, 'variables')
```
I don't like that we're still doing this, but I don't want to add one more deprecated thing into this PR.

But I might make an issue to get rid of all these instances, and wanted to ask how you would feel about that. In order to optionally take the "@" character, that means the parameter type is string, and I would much prefer that the type be a dict.

john-westcott-iv replied:
So I almost said I 100% agree on this but then I started to think about it....
The only reason this notation was supported was because it had to be fed into Tower CLI that way. Since we are deprecating Tower CLI with this PR one could argue that now is the perfect time to drop this. If we wanted to do that we could provide in the release notes something like:

IF you used the @ method to import a vars file, here is how you would convert your code to have Ansible load a vars file.... one of the caveats would be that the @ syntax as is loads the file from the targeted node where as using a lookup plugin would load from the controller.

Also, as an FYI, in my test I passed in something like:
   variables:
       foo: bar
Ansible generates a warning that its forcing a dict into a str but other than that it works.

AlanCoding replied:
> IF you used the @ method to import a vars file, here is how you would convert your code to have Ansible load a vars file....

probably similar to:

https://github.com/ansible/awx/blob/devel/awx_collection/plugins/modules/tower_job_template.py#L242

I'm cautiously optimistic that no one will have problems with these parameter type changes  - because the change was prompted by Ansible core itself. Their CI started enforcing types, but old modules were given an exception. The variables param here technically passes their CI, but doesn't respect the _spirit_ of enforcing types. So at minimum, we have someone else to point the finger at.

beeankha replied:
We can go ahead and take this out (changes currently in progress).  Do we want to raise an exception or `warn` that the "@" syntax is no longer supported?

AlanCoding commented on awx_collection/plugins/modules/tower_inventory.py:
```diff
@@ -106,31 +106,37 @@ def main():
+    elif state == 'present':
+        # We need to perform a check to make sure you are not trying to convert a regular inventory into a smart one.
+        if inventory and inventory['kind'] == '' and inventory_fields['kind'] == 'smart':
+            module.fail_json(msg='You cannot turn a regular inventory into a "smart" inventory.')
```
Won't the API itself give a 400 in this event? If it does that, and if the 400 error is handled correctly, then I don't think we should need this.

that case can be unit tested. As long as it gives a sensible error message that says mostly the same thing, we shouldn't need this conditional.

beeankha replied:
The API returns a `405` without a helpful error message:

![Screen Shot 2020-02-10 at 9 32 16 AM](https://user-images.githubusercontent.com/28930622/74158617-63f94d80-4be8-11ea-87fb-a6f9a35ac64a.png)

I filed a bug for this here: https://github.com/ansible/awx/issues/5744

AlanCoding commented on awx_collection/plugins/modules/tower_organization.py:
```diff
@@ -66,49 +77,61 @@
+        description=dict(type='str', required=False),
         custom_virtualenv=dict(type='str', required=False),
-        state=dict(choices=['present', 'absent'], default='present'),
+        max_hosts=dict(type='str', required=False, default="0"),
```
why not just make the type here integer and then get rid of the custom logic to enforce that?

john-westcott-iv replied:
We probably should do this. In an older release of Ansible I had problems getting an integer field into a module because it would find a string and die. However, it looks like this has been fixed somewhere along the line.

AlanCoding commented on awx_collection/plugins/modules/tower_team.py:
```diff
@@ -57,57 +67,53 @@
+        }
+    })
+
+    # Create data to sent to create and update
```
sent/send

...or delete comment?

john-westcott-iv replied:
Fixed.

AlanCoding commented on awx_collection/test/awx/conftest.py:
```diff
@@ -42,12 +42,16 @@ def sanitize_dict(din):
 
 @pytest.fixture
 def run_module(request):
+    # A placeholder to use while modules get converted
```
don't think you need this comment anymore?

AlanCoding commented on awx_collection/test/awx/test_project.py:
```diff
@@ -15,15 +15,19 @@ def test_create_project(run_module, admin_user, organization):
         'id': proj.id,
-        'project': 'foo',
-        'state': 'present'
+        'warnings': warning
```
👍

AlanCoding commented on awx_collection/plugins/modules/tower_credential_type.py:
```diff
@@ -127,53 +115,49 @@ def main():
+    # Add entries to json_output to match old module
+    module.json_output['credential_type'] = name
+    module.json_output['state'] = state
+    module.json_output['existing_credential_type'] = credential_type
```
Without some larger framework to deal with this, I would prefer to get rid of it. I think the ultimate solution would be diff mode, which is fully documented and people know how to use.

john-westcott-iv committed:
Adding mutually exclusive if functionality to support tower_inventory_source

AlanCoding commented on awx_collection/plugins/modules/tower_organization.py:
```diff
@@ -66,49 +77,56 @@
+    # Attempt to look up the related items the user specified (these will fail the module if not found)
+    # instance_group_objects = []
+    # for instance_name in instance_group_names:
+    #     instance_group_objects.append(module.resolve_name_to_id('instance_groups', instance_name))
```
We could file a new issue for org instance groups if there isn't one already. Let's get rid of the comments for getting this wrapped up.

AlanCoding commented on awx_collection/plugins/modules/tower_user.py:
```diff
@@ -106,59 +107,50 @@
+        'username': module.params.get('username'),
+        'first_name': module.params.get('first_name'),
+        'last_name': module.params.get('last_name'),
+        'password': module.params.get('password'),
```
how are null values handled?? Does the `create_or_update_if_needed` method ignore them? I'm interested in the case where the module used, and non-required parameters are not provided.

beeankha replied:
In the `update_if_needed` method in the new module utils, the code below should be taking care of this:

```
            for field in new_item:
                existing_field = existing_item.get(field, None)
                new_field = new_item.get(field, None)
                # If the two items don't match and we are not comparing '' to None
                if existing_field != new_field and not (existing_field in (None, '') and new_field == ''):
                    # Something doesn't match so let's update it
                    needs_update = True
                    break
```

At the module level, we should only be adding in the parameters if the optional ones are set (like what you did most recently in `inventory_source`).

john-westcott-iv committed:
Converting tower_inventory_source

Fix up inventory_source module changes, fix import yaml sanity error, change inventory_source unit tests to comply with new structure.

john-westcott-iv committed:
Converted tower_job_launch.py

john-westcott-iv committed:
Converted tower_job_cancel

john-westcott-iv committed:
Fixing config loading issues when the config has no [general] section

Fixed typo in help documentation

Fix up sanity errors and update converted modules

Remove unnecessary test playbook file

beeankha committed:
Fix Python2 config incompatibility issue

beeankha committed:
Attempt to make validate_certs work in Python2

AlanCoding commented:
Don't all the options in here:

https://github.com/ansible/awx/blob/ac50377001862baa9816058aeb9a79ba82e56192/awx_collection/plugins/doc_fragments/auth.py

need to optionally source from env???

like

https://github.com/ansible/awx/blob/c33d2a1e00abb137047d96d17d8763a8e0e7265d/awx_collection/plugins/inventory/tower.py#L34-L35

previously, tower-cli did this (I think)

beeankha committed:
Properly cast verify_ssl type to a bool

john-westcott-iv committed:
Updates to config file loading

Now supports json or yaml

Depricated multiple k=v on a single line

Remove assert statement and unused import from module_util

squidboylan committed:
Fix strtobool casting

AlanCoding commented on awx_collection/plugins/modules/tower_inventory_source.py:
```diff
@@ -20,333 +20,346 @@
+                             ('source', 'openstack', ['credential']),
+                             ('source', 'rhv', ['credential']),
+                             ('source', 'tower', ['credential']),
+                             ('source', 'custom', ['source_script']),
```
I'm okay applying these rules specific to scm versus cloud sources. But I would prefer to have a local variable which lists all the cloud sources in a list, like `['gce', 'azure_rm', ...]`, and then building this list from those. If there are rules unique to certain cloud sources, then that's going in a direction that I do not want to maintain.

AlanCoding commented on awx_collection/plugins/modules/tower_inventory_source.py:
```diff
@@ -20,333 +20,346 @@
+    # source_regions		?	?	-	o	o	o		-	-	-		-		-	-	-
+    # instance_filters		?	?	-	o	-	-		o	-	-		-		-	o	-
+    # group_by			?	?	-	o	-	-		o	-	-		-		-	-	-
+    # source_vars*		?	?	-	o	-	o		o	o	o		o		-	-	-
```
If regions, filters, group_by, or source_vars are accepted by the API, but not functional for that particular inventory source type, then let's file it as an API bug.

john-westcott-iv replied:
We have a task to file a bug for the inventory_sources in general. Try doing a POST to api/v2/inventory_sources with the payload:
```
{
"name": "Some Fubar Source",
"source": "scm",
"inventory": 1
}
```

 spolier alert:
![image](https://user-images.githubusercontent.com/32551173/74346512-b1f38a00-4d7d-11ea-8a67-5d9f3c72eb94.png)

Based on the epic state of this failure I didn't even bother trying all of the various other combinations and just put a little elbow grease into the module to try and enforce proper state.

beeankha committed:
Update inventory_source module source_script parameter to be optional

Unitied comment

Fix up inventory_source example, misc comment edits

AlanCoding commented on awx_collection/plugins/modules/tower_inventory_source.py:
```diff
@@ -20,333 +20,347 @@
+  tower_group:
+    name: localhost
+    description: "Local Host Group"
+    inventory: "Local Inventory"
```
What's going on with this example? This is the `tower_inventory_source` module, so the example should be about that and not `tower_group`.

beeankha replied:
Woops!  Will update that example.

AlanCoding commented on awx_collection/plugins/modules/tower_inventory_source.py:
```diff
@@ -20,333 +20,347 @@
+    optional_vars['update_on_launch'] = module.params.get('update_on_launch')
+    optional_vars['update_cache_timeout'] = module.params.get('update_cache_timeout')
+    source_project = module.params.get('source_project')
+    optional_vars['update_on_project_update'] = module.params.get('update_on_project_update')
```
On the principle of consistency, I'm not liking this `option_vars` dict. If I read it right, the difference between it and the `*_fields` dict is that it holds falsy values. That's not something that we need, I'd rather just test falsyness before we put it into the `inventory_source_fields`

Again, the thing I really want to double-down on here is that the modules should all look very consistent, and this is one I really want to bring into line. Otherwise we lose out on the benefits from a lot of the rest of the work.

john-westcott-iv replied:
For this one, this was kind of a "hack" because of the issues with the API in regards to inventory sources. The only thing we know we will have to pass back to the API is in the lines:
```
    inventory_source_fields = {
        'name': new_name if new_name else name,
        'inventory': inventory_id,
    }
```

All of the other parameters may or may not be present based on the inventory source type. So, instead of doing a bunch of
```
if <param> is not None:
    inventory_source_fields[<param>] = <param_varaibale>
```

I threw everything into the `optional_vars` dict and then down below:
```
    # Layer in all remaining optional information
    for field_name in optional_vars:
        if optional_vars[field_name]:
            inventory_source_fields[field_name] = optional_vars[field_name]
```

We have on the list to file a bug against the API but until that is resolved this just saves us a bunch of lines of code.

With that said, I see your point about the patten we could add the additional lines instead of the dict /w for.

AlanCoding replied:
What bug would you file against the API? The canonical behavior with regards to Django and DRF is that CharField type will not accept null values, you have to use a null-char-field, which is a very recent construct, and we're not using that for these fields. Several of the fields should be blank under normal conditions. But handling `None` is something that all modules should be obligated to do - because tower-cli did this. It popped the `None` values out of the dict before it used it in data to the API. It is true that some fields are nullable, but we honestly could never really well handle the act of nulling it (I don't expect that we will start).

So if non-provided values are actually passed to the API as null, that could be an issue for _all the other modules_ because when you mention "doing a bunch of" checking if the params are null, that's something I expect to see. Now we can clean the code up, just loop over a tuple of the field names or something, so an extra check per-field doesn't requires a ton more code.

I hope I'm making sense.

john-westcott-iv committed:
Fixed issue that caused warning message to always display

Because scm_update_cache_timeout has a default and thus will always be != None

john-westcott-iv committed:
Only try an ID lookup if we can convert the field name_or_id to an integer

Fix linting issues, update tower_project unit test

AlanCoding commented on awx_collection/plugins/modules/tower_credential_type.py:
```diff
@@ -127,53 +115,48 @@ def main():
-    json_output['changed'] = result['changed']
-    module.exit_json(**json_output)
+    # Deal with check mode
+    module.default_check_mode()
```
Should this go away because it'll be caught in subsequent requests?

john-westcott-iv committed:
Remove reference to default check mode

john-westcott-iv committed:
Removing manual and file source types and correcting default for custom_virtualenv

john-westcott-iv committed:
Making variables work for hosts

Clear up sanity test and remove redundant import statement

beeankha commented:
recheck

AlanCoding commented:
Just rebased https://github.com/john-westcott-iv/awx/pull/3 to address unit test issues.

AlanCoding committed:
Optional tower cli (#3)

* Allow running tests without tower_cli

* patch up test mutability

* Fix test import error, warning mock

* flake8 error

Update documentation for non-converted modules

AlanCoding commented on awx_collection/plugins/modules/tower_credential_type.py:
```diff
@@ -127,53 +115,45 @@ def main():
+
+    # Add entries to json_output to match old module
+    module.json_output['credential_type'] = name
+    module.json_output['state'] = state
```
Wouldn't we prefer `state` to be part of the returned data? If you run a task in check mode, then you might make decisions based on what the state is, and it would make sense to report this even if it was given as a parameter. In that case, we should lose this line and have `delete_if_needed` and `create_or_update_if_needed` apply the state field.

AlanCoding commented on awx_collection/test/awx/test_credential.py:
```diff
@@ -78,13 +78,13 @@ def test_create_custom_credential_type(run_module, admin_user):
     result.pop('invocation')
+    result.pop('name')
     assert result == {
         "credential_type": "Nexus",
```
No one is going to look at this and know why "name" was popped off without an assertion, but "credential_type" is in the assertion data.

I would swap the two and add a comment that "credential_type" is only provided in the return data for backward compatibility.

This is important if we declare the RETURNS documentation later.

AlanCoding commented on awx_collection/test/awx/test_group.py:
```diff
@@ -26,9 +26,8 @@ def test_create_group(run_module, admin_user):
     assert result == {
         'id': group.id,
-        'group': 'Test Group',
+        'name': 'Test Group',
```
Why is "credential_type" provided for backward compatibility, but "group" is not?

I'm not against dropping support for a returned key, but it needs to be declared in the README.md with the other backward-incompatible changes, and one way or the other, these should be consistent about whether they kept the old ones or got rid of them.

AlanCoding commented on awx_collection/test/awx/test_group.py:
```diff
@@ -53,7 +52,5 @@ def test_tower_group_idempotent(run_module, admin_user):
         'id': group.id,
-        'group': 'Test Group',
         'changed': False,  # idempotency assertion
-        'state': 'present'
```
Is there any reason for having "name" in the returned data when creating, but not when modifying an object which already exists? That looks inconsistent. If you are including keys because they were non-null params, then okay, but that's not a consistent explanation either, because "name" _was_ provided in the invocation.

AlanCoding commented:
Sanity might be impacted by https://github.com/ansible/ansible/pull/66386

AlanCoding committed:
[last PR stuff] + Add warning if configs specified in 2 params (#5)

* Lean on API validation for tower_inventory_source arg errors

used for
 - validating needed credential is given
 - missing source_project for scm sources

* Add warning when config is specified in 2 places

Fix up unit tests, address multiple comments re: backwards compatibility, redundant methods, etc.

Update new_name and variables parameters, update unit tests

beeankha committed:
Change default values for dict parameter

Removing default of empty dict from variables param on group and host modules

Make modules comply with updated sanity tests

AlanCoding commented on awx_collection/plugins/modules/tower_group.py:
```diff
@@ -82,7 +82,7 @@ def main():
         description=dict(),
         inventory=dict(required=True),
-        variables=dict(type='dict', required=False),
+        variables=dict(type='dict', required=False, default={}),
```
What does this mean for the case where a group exists with variables, and then a task tries to change some other parameter?

I'm comfortable with getting `None` when a parameter is not provided. NoneType isn't a dict or a string, so it doesn't mix up not-provided with false-y values. If the default is an empty dict, the best case scenario is that it ignores empty dict when the user gives it, and the worse case scenario is that it deletes variables when the user does unrelated stuff.

beeankha replied:
Per our offline discussion, I will remove `default={}` from this parameter.

beeankha commented:
Latest changes comply with the updated sanity tests from this PR: https://github.com/ansible/ansible/pull/66386

AlanCoding commented on awx_collection/plugins/modules/tower_project.py:
```diff
@@ -133,107 +148,128 @@
+    if scm_credential is not None:
+        project_fields['credential'] = scm_credential_id
+    if scm_allow_override is not None:
+        project_fields['scm_allow_override'] = scm_allow_override
```
Why are some of these going through the `is not None` check but others aren't?

beeankha replied:
I've updated the other non-required params in the modules to go through the same check; turns out this affected idempotency behavior.

AlanCoding commented on awx_collection/plugins/modules/tower_group.py:
```diff
@@ -28,6 +28,11 @@
+      description:
+        - A new name for this group (for renaming)
+      required: False
+      type: str
```
All of the new params should have a `version_added` in this documentation. I think we want to go with 3.7 for that, indicating the Tower version it would work with. Even though this is confusing.

AlanCoding reviewed (approved):
We went over a last batch of changes with
 - added entry to backward-incompatible changes about return types
 - change type of tower_inventory variables to match host and group
 - added some version_added
 - remove empty string default for some description fields

With those in, I say we're good to go.

beeankha committed:
Update documentation for modules, add deprecation warning for role, send and receive modules

Update variables in tower_inventory to be in dict format

beeankha commented:
recheck

AlanCoding commented:
recheck

AlanCoding commented:
Could you add the deprecated flag to the workflow module's schema parameter? That will be replaced by import/export. Or possibly partially replaced by new modules.

beeankha committed:
Make non-required params actually optional, fix idempotency issues
